### PR TITLE
CA-287863: xe vm-shutdown complete the task too early

### PR DESCRIPTION
**Pull request by krizex**
When master forward a task to slave, slave will mark the task as completed once
finish processing it, however when master finishing the forwarding, it will also
mark the task as completed. Furthermore, when slave finished the task, that does
not mean the task is done, there may be some stuff that master has to handle, so
we have to ensure the task only complete once the master finish its work.

Signed-off-by: Yang Qian <yang.qian@citrix.com>

### CODE REVIEW

File: ocaml/xapi/server_helpers.ml [hunk 1]
```diff
@@ -75,6 +75,8 @@ let exec ?marshaller ?f_forward ~__context f =
           (* use the forwarding layer (NB this might make a local call ultimately) *)
           forward ~local_fn:f ~__context
     in
+    (* For forwarded task, we should not complete it here, the server which forward the task will complete it *)
+    if not (Context.forwarded_task __context) then
```
Comment (lindig) [inline ocaml/xapi/server_helpers.ml]: This code has become dependent on a condition but indentation has not changed to reflect this. Please check the indentation of the code.

**Comment by robhoes** [comment]
Don't merge yet though, because some BST tests seem to be failing. Please check those out first.

File: ocaml/xapi/server_helpers.ml [hunk 2]
```diff
@@ -75,19 +75,21 @@ let exec ?marshaller ?f_forward ~__context f =
           (* use the forwarding layer (NB this might make a local call ultimately) *)
           forward ~local_fn:f ~__context
     in
-    begin match marshaller with
+    (* For forwarded task, we should not complete it here, the server which forward the task will complete it *)
+    if not (Context.forwarded_task __context) then begin
+      match marshaller with
       | None    -> TaskHelper.complete ~__context None
       | Some fn -> TaskHelper.complete ~__context (Some (fn result))
```
Comment (krizex) [inline ocaml/xapi/server_helpers.ml]: @robhoes I have some concern about this. After this change, the result in the slave won't be updated to the task result, but instead will update with the execution result in the master, will it bring some other issues?

Comment (robhoes) [reply ocaml/xapi/server_helpers.ml]: I should hope that the result of the forwarded call is already always returned (unaltered) by the master to the original caller, especially for synchronous calls. So I think this is fine.

**Comment by robhoes** [comment]
The failing test seems to be an import that times out. We should check how the task completion is handled for such calls, which don't come as RPCs through server.ml, but have their own HTTP handlers.

**Comment by krizex** [comment]
Seems the installation of the build is failed, so all BST & BVT failed. I have updated with the latest specs of team/ring3/master and has started rerun the test again.
Privately I have tested this fix by uploading xapi to my hosts and the error messages gone.

**Comment by robhoes** [comment]
Well, the installation was OK, but the 62-create-guest-templates firstboot script failed:
```
Load /usr/share/xapi/vm-templates/sles-11-sp4-64bit.json
Load /usr/share/xapi/vm-templates/ubuntu-12.04-64bit.json
Load /usr/share/xapi/vm-templates/base-windows-8.json
Insert 11fd3dc9-96cc-49af-b091-a2ca7e94c589
Traceback (most recent call last):
  File "/usr/bin/create-guest-templates", line 17, in <module>
    loader.insert_templates()
  File "/usr/lib/python2.7/site-packages/guesttemplates/loader.py", line 181, in insert_templates
    self._insert_template(i)
  File "/usr/lib/python2.7/site-packages/guesttemplates/loader.py", line 165, in _insert_template
    raise RuntimeError("Template import timeout")
RuntimeError: Template import timeout
```
This does a VM import.

**Comment by krizex** [comment]
Hmm, I will take a look at that tomorrow.

**Comment by robhoes** [comment]
Hmm... I wonder why `exec_with_forwarded_task` is used in `Xapi_http.with_context`. That looks odd to me. In this patch, I don't quite like the new `~with_task` param. I'll look a bit further...

**Comment by krizex** [comment]
@robhoes The name of `has_task` comes from the original comment https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/server_helpers.ml#L59
I meant to introduce a new param `~need_complete` ;-)

**Comment by robhoes** [comment]
Ah, that was another mysterious comment indeed :)

**Comment by robhoes** [comment]
I think that it is declared a forwarded task, to make sure that the task is not destroyed when the operation completes; the task is created by the client, and should be destroyed by the client.

I also notice that the "forwarded task destroyed" log lines that we often see actually mean that the forwarded task was _not_ destroyed...

**Comment by robhoes** [comment]
Well that is indeed the problem and one of the sources of confusion: the word forwarded is used to different things, or at least the function is "abused" for different things. I think that we should reserve the term "forwarding" for the action of passing a call from the master to a slave (as in "message_forwarding").

I have some ideas now to simplify some of this area, and will do that in follow-up PRs. I'll accept this PR now, because it already fixes a real issue.

**Pull request merged**